### ISSUE TITLE
training: standalone offline pretraining -> checkpoint manifest -> online fine-tuning lane (#4012 deferred maintainer request)

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#4245]
## Goal
PR #4170 delivered the SAC replay warm-start smoke but explicitly skipped the maintainer's late request: a STANDALONE offline pretraining stage with durable checkpoint provenance. This is the missing precondition for SAC's entry into the 7-arm comparison matrix (#4244).

## Acceptance criteria (per the #4170 review record)
1. `pretrain_offline_policy.py`: offline pretraining entry point over the RL trajectory dataset pipeline (#4145 lineage); emits checkpoint + normalizer + manifest (dataset identity, steps, hashes).
2. `offline_to_online_finetune.py`: loads that checkpoint via its manifest and fine-tunes online; provenance chain recorded end to end.
3. CPU smoke for both stages (tiny budgets); compact evidence under docs/context/evidence/.
4. Comparison vs scratch SAC only AFTER dataset/checkpoint provenance is complete — and only as a predeclared arm in #4244, not ad hoc.

Parent: #4012 (stays open as the lane tracker).

**Comment by ll7** [comment]
## Implementation plan

Scope: implement the deferred standalone offline-pretraining → durable checkpoint manifest → online fine-tuning lane for #4012. This is an infrastructure/provenance lane for SAC/offline-online RL, not a benchmark-performance claim and not the #4244 comparison itself.

### 1. Preserve the #4170 boundary and reuse its primitives

PR #4170 already delivered:

- `RLTrajectoryDataset.v1` train-split loading;
- strict observation/action preflight against the online SAC environment;
- replay-buffer seeding from validated offline transitions;
- optional offline gradient warm-up before online SAC continuation;
- a diagnostic scratch comparator wrapper;
- local ignored smoke artifacts only.

This issue should not duplicate those primitives. It should add the missing standalone checkpoint-producing stage and a manifest-driven fine-tuning stage.

Implementation anchors:

```text
robot_sf/training/offline_online_rl.py
robot_sf/training/hybrid_replay_buffer.py
scripts/training/train_sac_sb3.py
scripts/training/run_offline_online_rl.py
configs/training/offline_online_rl/
tests/training/test_offline_online_rl.py
tests/training/test_train_sac_sb3_offline_online.py
tests/training/test_run_offline_online_rl.py
```

### 2. Define a checkpoint manifest contract

Add a typed contract module or dataclass, for example:

```text
robot_sf/training/offline_pretraining_manifest.py
```

Manifest schema:

```text
offline-policy-checkpoint-manifest.v1
```

Recommended fields:

```json
{
  "schema_version": "offline-policy-checkpoint-manifest.v1",
  "issue": 4245,
  "parent_issue": 4012,
  "claim_boundary": "standalone offline pretraining provenance only; not benchmark evidence",
  "algorithm": "sac",
  "policy_type": "MlpPolicy",
  "checkpoint_path": "output/.../offline_pretrain_checkpoint.zip",
  "normalizer_path": "output/.../vecnormalize.pkl",
  "checkpoint_sha256": "...",
  "normalizer_sha256": "...",
  "training_config_path": "configs/training/offline_online_rl/issue_4245_pretrain_smoke.yaml",
  "training_config_sha256": "...",
  "dataset": {
    "dataset_path": "...",
    "dataset_manifest_path": "...",
    "dataset_sha256": "...",
    "dataset_schema_version": "RLTrajectoryDataset.v1",
    "dataset_id": "...",
    "split": "train",
    "episode_count": 1,
    "accepted_transitions": 2,
    "dropped_terminal_transitions": 0
  },
  "offline_training": {
    "gradient_steps": 1,
    "batch_size": 8,
    "seed": 4245,
    "observation_contract": "...",
    "action_contract": "..."
  },
  "environment_contract": {
    "scenario_config": "...",
    "observation_space_fingerprint": "...",
    "action_space_fingerprint": "..."
  },
  "created_at_utc": "...",
  "git_commit": "..."
}
```

Rules:

- Manifest must be complete before online fine-tuning accepts the checkpoint.
- Missing checkpoint, normalizer, dataset manifest, SHA-256, or env-space fingerprint fails closed.
- Manifest paths may be local output paths for smoke, but compact evidence must record hashes and claim boundary.
- Do not place model checkpoint bytes under `docs/context/evidence/`.

### 3. Add standalone offline pretraining entry point

Add:

```text
scripts/training/pretrain_offline_policy.py
```

CLI sketch:

```bash
uv run --extra training python scripts/training/pretrain_offline_policy.py \
  --config configs/training/offline_online_rl/issue_4245_pretrain_smoke.yaml \
  --output-dir output/issue_4245_offline_pretrain_smoke \
  --manifest-out output/issue_4245_offline_pretrain_smoke/offline_checkpoint_manifest.json
```

Responsibilities:

1. Load SAC/offline config using existing config loader patterns.
2. Load and validate `RLTrajectoryDataset.v1` via `load_offline_transition_batch(...)`.
3. Validate observations/actions against the online SAC env spaces using existing `validate_batch_against_env_spaces(...)`.
4. Instantiate SAC and seed replay buffer or run offline-only gradient updates without online environment stepping.
5. Save checkpoint and normalizer/VecNormalize state when present.
6. Write `offline-policy-checkpoint-manifest.v1` with dataset identity, hashes, steps, action/observation contract, and output checksums.
7. Exit nonzero if any provenance field cannot be populated.

Keep CPU smoke budgets tiny. The point is durable provenance wiring, not policy quality.

### 4. Add manifest-driven online fine-tuning entry point

Add:

```text
scripts/training/offline_to_online_finetune.py
```

CLI sketch:

```bash
uv run --extra training python scripts/training/offline_to_online_finetune.py \
  --config configs/training/offline_online_rl/issue_4245_finetune_smoke.yaml \
  --pretrained-manifest output/issue_4245_offline_pretrain_smoke/offline_checkpoint_manifest.json \
  --output-dir output/issue_4245_offline_to_online_finetune_smoke \
  --manifest-out output/issue_4245_offline_to_online_finetune_smoke/finetune_manifest.json
```

Responsibilities:

1. Load and validate the offline checkpoint manifest.
2. Verify checkpoint and normalizer hashes before loading.
3. Rebuild online environment and assert observation/action-space fingerprints match or are explicitly compatible.
4. Load the SAC checkpoint and continue online training for a tiny budget.
5. Emit a new manifest:

```text
offline-to-online-finetune-manifest.v1
```

Manifest should include:

- parent offline checkpoint manifest path and SHA-256;
- inherited dataset identity;
- loaded checkpoint SHA-256;
- online fine-tune config and SHA-256;
- online timesteps;
- output checkpoint and normalizer hashes;
- seed;
- environment/scenario config;
- claim boundary: diagnostic smoke/provenance only.

### 5. Add configs for the smoke lane

Add checked-in CPU smoke configs:

```text
configs/training/offline_online_rl/issue_4245_pretrain_smoke.yaml
configs/training/offline_online_rl/issue_4245_finetune_smoke.yaml
```

Optional wrapper config:

```text
configs/training/offline_online_rl/issue_4245_offline_pretrain_to_finetune_smoke.yaml
```

Config rules:

- tiny dataset/smoke path under ignored `output/`;
- CPU-only SB3 parameters;
- deterministic seed;
- minimal gradient steps / timesteps;
- explicit scenario config;
- explicit dataset manifest path;
- no scratch comparator inside this issue except fixture tests proving manifest compatibility. Scratch comparison belongs to #4244 only after provenance is complete.

### 6. Add compact evidence under docs/context/evidence

Add a compact evidence packet after smoke runs:

```text
docs/context/evidence/issue_4245_offline_pretrain_finetune_smoke/
```

Recommended files:

```text
README.md
pretrain_manifest_summary.json
finetune_manifest_summary.json
provenance_chain.json
smoke_report.md
SHA256SUMS
```

Do not commit:

- `.zip` model checkpoints;
- normalizer pickle files;
- replay buffers;
- raw generated datasets beyond tiny fixture data already intentionally tracked elsewhere;
- TensorBoard logs;
- W&B state;
- Slurm logs.

The compact summaries should include hashes and local output pointers, not model bytes. `README.md` should explicitly state that these artifacts prove lane/provenance integrity only.

### 7. Add tests

Extend existing tests and add new focused tests:

```text
tests/training/test_offline_pretraining_manifest.py
tests/training/test_pretrain_offline_policy.py
tests/training/test_offline_to_online_finetune.py
```

Minimum tests:

- manifest validation accepts a complete fixture and rejects missing checkpoint path;
- manifest validation rejects checksum mismatch;
- manifest validation rejects missing dataset identity or dataset SHA;
- offline pretrain CLI writes checkpoint, normalizer if present, and manifest for a tiny fixture;
- online finetune CLI refuses a missing or mismatched pretrained manifest;
- online finetune records parent manifest SHA and output checkpoint SHA;
- environment/action-space mismatch fails closed;
- no raw checkpoint bytes are copied into `docs/context/evidence/`;
- generated compact evidence contains `eligible_for_claim=false` or equivalent claim boundary;
- scratch-comparator execution remains out of this lane unless #4244 calls it via a predeclared arm.

### 8. Optional orchestration wrapper

If useful, add a wrapper that runs both stages sequentially:

```text
scripts/training/run_offline_pretrain_to_online_finetune.py
```

This wrapper should only orchestrate:

1. `pretrain_offline_policy.py`;
2. `offline_to_online_finetune.py` with the manifest from step 1;
3. compact summary generation.

It should not introduce additional training semantics.

### 9. Validation commands

```bash
uv run --extra training pytest \
  tests/training/test_offline_online_rl.py \
  tests/training/test_hybrid_replay_buffer.py \
  tests/training/test_train_sac_sb3_offline_online.py \
  tests/training/test_offline_pretraining_manifest.py \
  tests/training/test_pretrain_offline_policy.py \
  tests/training/test_offline_to_online_finetune.py -q

uv run ruff check \
  robot_sf/training/offline_pretraining_manifest.py \
  scripts/training/pretrain_offline_policy.py \
  scripts/training/offline_to_online_finetune.py \
  tests/training/test_offline_pretraining_manifest.py \
  tests/training/test_pretrain_offline_policy.py \
  tests/training/test_offline_to_online_finetune.py

uv run --extra training python scripts/training/pretrain_offline_policy.py \
  --config configs/training/offline_online_rl/issue_4245_pretrain_smoke.yaml \
  --output-dir output/issue_4245_offline_pretrain_smoke \
  --manifest-out output/issue_4245_offline_pretrain_smoke/offline_checkpoint_manifest.json

uv run --extra training python scripts/training/offline_to_online_finetune.py \
  --config configs/training/offline_online_rl/issue_4245_finetune_smoke.yaml \
  --pretrained-manifest output/issue_4245_offline_pretrain_smoke/offline_checkpoint_manifest.json \
  --output-dir output/issue_4245_offline_to_online_finetune_smoke \
  --manifest-out output/issue_4245_offline_to_online_finetune_smoke/finetune_manifest.json
```

### 10. Handoff to #4244

After this issue is complete, #4244 may add a predeclared comparison arm. The handoff surface should be a stable manifest path and config reference, for example:

```yaml
pretrained_sac_arm:
  source_issue: 4245
  offline_checkpoint_manifest: output/issue_4245_offline_pretrain_smoke/offline_checkpoint_manifest.json
  finetune_manifest: output/issue_4245_offline_to_online_finetune_smoke/finetune_manifest.json
  eligibility: provenance_complete_diagnostic_only
```

#4244 must not compare against scratch SAC until this provenance chain is present and reviewed.

### Definition of done

- `pretrain_offline_policy.py` exists and emits checkpoint, normalizer if applicable, and `offline-policy-checkpoint-manifest.v1`.
- `offline_to_online_finetune.py` loads that manifest, verifies hashes, fine-tunes online, and emits a chained manifest.
- CPU smoke for both stages passes with tiny budgets.
- Compact evidence under `docs/context/evidence/` records the provenance chain without checkpoint bytes.
- Tests cover manifest completeness, hash mismatch, env-space mismatch, CLI smoke, and evidence artifact safety.
- No benchmark/performance claim, scratch comparison, paper/dissertation claim, or #4244 matrix inclusion is made inside this issue.
